File: 2024/Day20/main.py
import numpy as np

SAMPLE_DATA: bool = False
if SAMPLE_DATA:
    filename = "sample_data.txt"
else:
    filename = "data.txt"


with open(filename, 'r') as f:
    data = [line.strip() for line in f]

maze = np.array([[c for c in l] for l in data], dtype=str)

rows, cols = maze.shape

def is_valid(x: int, y: int) -> bool:
    return  x >= 0 and x < rows and y>=0 and y < cols and maze[x, y] != '#'

def manhattan(a: tuple[int, int], b: tuple[int, int]) -> int:
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

offsets = [ (0,1),(-1,0), (0,-1), (1,0) ]

def find_path(start: tuple[int, int], end: tuple[int, int]):
    distances = {}
    current = end
    target = end
    count = 0
    while start != target:
        distances[target] = count
        count += 1
        for (xoff, yoff) in offsets:
            neighbor = (target[0]+xoff, target[1]+yoff)
            if neighbor != current and is_valid(*neighbor):
                current = target 
                target = neighbor
                break
    distances[start] = count
    return distances

start = tuple(np.argwhere(maze == 'S')[0])
end = tuple(np.argwhere(maze == 'E')[0])

distances = find_path(start, end)

part1 = 0 
part2 = 0
for p1, c1 in reversed(distances.items()):
    for p2, c2 in distances.items():
        dist = manhattan(p1, p2) 
        if dist <= 2 and c2 - c1 - dist >= 100:
            part1 += 1
        if dist <= 20 and c2 - c1 - dist >= 100:
            part2 += 1
print(f'{part1} {part2}')

File: 2024/Day20/main.cpp
#include <algorithm>
#include <charconv>
#include <fstream>
#include <ranges>
#include <string>
#include <string_view>
#include <unordered_map>
#include <vector>
#include <cstdio>
#include <numeric>
#include <print>
#include <cinttypes>
#include <utility>

namespace r = std::ranges;
namespace v = std::ranges::views;
using namespace std::string_view_literals;

constexpr bool SAMPLE_DATA = false;

std::vector<std::string> read_file(std::string_view filename) {
    std::ifstream file(filename.data());
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

using matrix = std::vector<std::string>;
using point = std::pair<int64_t, int64_t>;

template <>
struct std::hash<point> {
    constexpr static inline std::hash<int64_t> hasher{};
    std::size_t operator()(const point& dir) const {
        const auto& [x, y] = dir;
        auto res = 0x345678L;
        int64_t mult = 1000003L;

        res = (res ^ hasher(x)) * mult;
        mult += 69068L + 4;

        res = (res ^ hasher(y)) * mult;
        mult += 69068L + 2;

        return res;
    }
};

point find_value(const matrix& m, char value) {
    for (size_t i = 0; i < m.size(); ++i) {
        for (size_t j = 0; j < m[i].size(); ++j) {
            if (m[i][j] == value) {
                return {i, j};
            }
        }
    }
    return {-1, -1};
}

bool is_valid(const matrix& m, int64_t row, int64_t col) {
    return row >= 0 && row < m.size() && col >= 0 && col < m[0].size() && m[row][col] != '#';
}

int64_t manhattan(const point& a, const point& b) {
    return std::abs(a.first - b.first) + std::abs(a.second - b.second);
}

static constexpr inline std::array offsets{
    point{0, 1},
    point{0, -1},
    point{1, 0},
    point{-1, 0}
};

std::unordered_map<point, int64_t> find_path(const matrix& m, point start, point end) {
    std::unordered_map<point, int64_t> distances{};
    auto current = end;
    auto target = end;
    int64_t count = 0;
    while (start != target) {
        distances[target] = count;
        count += 1;
        for (auto [xoff, yoff] : offsets) {
            auto neighbor = point{target.first + xoff, target.second + yoff};
            if (neighbor != current && is_valid(m, neighbor.first, neighbor.second)) {
                current = target;
                target = neighbor;
                break;
            }
        }
    }
    distances[start] = count;
    return distances;
}

int main() {
    auto lines = read_file(SAMPLE_DATA ? "sample_data.txt"sv : "data.txt"sv);
    auto start_tile = find_value(lines, 'S');
    auto end_tile = find_value(lines, 'E');
    auto distances = find_path(lines, start_tile, end_tile);

    auto part1 = 0;
    auto part2 = 0;

    for (const auto& [p1, c1] : distances | v::reverse) {
        for (const auto& [p2, c2] : distances) {
            auto dist = manhattan(p1, p2);
            if (dist <= 2 && c2 - c1 - dist >= 100) {
                part1++;
            }
            if (dist <= 20 && c2 - c1 - dist >= 100) {
                part2++;
            }
        }
    }

    std::println("{} {}", part1, part2);
}